### ISSUE TITLE
Assigning issue to self doesn't assign it

### ISSUE THREAD

**Issue by h0adp0re** [ankitpokhrel/jira-cli#198]
**Describe the bug**

Running the following command doesn't assign the issue to me, instead the user search opens up, in which I'm unfindable as well.
```sh
jira issue assign "TICKET-CODE" "$(jira me)"
```
**Please provide following details**
1. JiraCLI Version:
    ```
    (Version="0.1.1", GitCommit="6663763b44a2e5d328b551ae7d2f1a37d1e232ae", GoVersion="go1.17.2", BuildDate="2021-11-03"", Compiler="gc", Platform="darwin/amd64")
    ```
2. Are you using Jira cloud or on-premise jira server? Also mention the version for on-premise installation.
    ```
    Jira Server v8.20.0
    ```
3. What operating system are you using? Also mention version.
    ```
    macOS 12.0.1
    ```
4. What terminal are you using? Also mention version.
    ```
    alacritty 0.9.0 (fed349a)
    ```

**To Reproduce**

1. Run `jira issue assign "TICKET-CODE" "$(jira me)"`
2. Issue isn't assigned to you, user search is opened up
3. Probably a separate issue but no names past the initial cropped list are findable

**Expected behavior**

The issue would be assigned to me.

**Additional context**

Not that it's relevant here but I'm using the line in the following function:
```sh
# Move an issue back to progress
jmbk() {
  jira issue transition "$1" "Back To Progress"
  jira issue assign "$1" "$(jira me)"
}
```

**Comment by ankitpokhrel** [comment]
I am not able to reproduce this but this can happen in 2 cases.
1. You have 2 users with the similar email/username and the tool cannot decide which one to use so it opens user search.
2. The user you are trying to assign is not assignable to the project. Are you able to assign this user from the UI? Alternatively, you can also check if the JiraAPI returns you this user as an assignable user.
   ```
   curl --request GET \
     --url 'https://<server>.example.com/rest/api/2/user/assignable/search?project=<KEY>' \
     --user "<username>:<password>"
   ```

Regarding `Probably a separate issue but no names past the initial cropped list are findable`, you can use the search feature to look for the specific user.

![search-assignee](https://user-images.githubusercontent.com/2364546/141976422-13fb00de-9fbf-4f3a-a660-d2c8d9d4c5a1.jpg)

**Comment by h0adp0re** [comment]
I am able to assign issues to myself via the UI, yes.
The command you suggested only returns users up to the letter `C`. We have very many employees and I see only a fraction of them in the response there.

> Regarding Probably a separate issue but no names past the initial cropped list are findable, you can use the search feature to look for the specific user.

This doesn't work in the same way.

**Comment by ankitpokhrel** [comment]
Just checked on a setup with 2000+ users (jira cloud), still cannot reproduce the issue :/

> The command you suggested only returns users up to the letter "C".

You can narrow the search using `query` parameter. For example:
```sh
curl --request GET \
  --url 'https://<server>.example.com/rest/api/2/user/assignable/search?project=<KEY>&query=<keyword>' \
  --user "<username>:<password>"
```

Also, does it work if you provide your name as displayed in JIra instead of doing `$(jira me)`? For instance:

```
jira issue assign "TICKET-CODE" "Your name as displayed in jira"
```

> The search doesn't work, as in it just returns to the list when I search for myself.

The search returns entries matching the keyword at the top of the list. Like in the screenshot below I am searching for the text `ankit`.

<img width="595" alt="Screenshot 2021-11-16 at 14 08 23" src="https://user-images.githubusercontent.com/2364546/141991000-f1e661d1-4a26-4d4e-8cee-4d3dd0855c6d.png">

Are you getting a different behavior? (If possible, screenshot or video of the issue might help)

**Comment by h0adp0re** [comment]
> You can narrow the search using query parameter.

The output is the same as before.

> Also, does it work if you provide your name as displayed in JIra instead of doing $(jira me)?

Nope, same result.

> The search returns entries matching the keyword at the top of the list. Like in the screenshot below I am searching for the text ankit.

Doesn't work like that for me. Even after supplying a keyword, the names in the "results" list remain unchanged. No filtering of any kind happens. Even when I type in the first name I see in the list.

Well, I don't know what to do here. I'm keeping my name private so sadly I won't be posting any visuals either.

**Comment by gennaro-tedesco** [comment]
I am seeing a similar behaviour myself too, in particular with some users that I do find via web interface but that do not show up in the list of possible users to assign to a ticket. Is it possible that the results of the API calls to JIRA are paginated and accidentally some users are being cut off from the pages (I have seen it happen with random users of different projects, so I could not derive any particular pattern in terms of access credentials or so)?

**Comment by ankitpokhrel** [comment]
>  Is it possible that the results of the API calls to JIRA are paginated and accidentally some users are being cut off from the pages

@gennaro-tedesco Yes the results are paginated and the tool fetches 100 results by default. There is a search feature to narrow down the results. Can you try steps mentioned in [the comment above](https://github.com/ankitpokhrel/jira-cli/issues/198#issuecomment-970265916) and see if you notice anything useful? Or, is there anything special about the users that are not showing up 🤔? Unfortunately, I am still not able to reproduce this issue.

**Comment by gennaro-tedesco** [comment]
So, strangely enough I have observed the following: let us assume we cannot find user X and that such user _is assignable to the project_ via JIRA UI.

- by querying the API directly via `curl` I do not get user X as available
- by replicating everything in the shoes of user X (namely user X uses your application with their own credentials) they _are_ available in the list of assignable users

so it seems to me that the JIRA API returns results in different orders (or does not return some at all) according to who is invoking the API call. I remember seeing a similar behaviour in a [project of mine](https://github.com/gennaro-tedesco/stargazer/blob/b5bc9b7409a3ebe4126c2b93aae343385f614e41/cmd/getters.go#L98) when using the GitHub API, that one has to explicitly indicate how many results per page to return otherwise it may be inconsistent. This is however speculation, it may be worth looking into a way to explicitly force pagination (if that even exists for JIRA).

**Comment by klaudiusz223** [comment]
Hi, 
I found 2 or even 3 different issues here.
1st - using displayName instead of name
if I change
https://github.com/ankitpokhrel/jira-cli/blob/5429cbd9d36360ac88245a412d3672a1c023a2a9/pkg/jira/types.go#L120-L126
to 
```
type User struct {
	AccountID   string `json:"accountId"`
	Email       string `json:"emailAddress"`
	DisplayName string `json:"displayName"`
	Name        string `json:"name"`
	Active      bool   `json:"active"`
}
```
I can successfully assign issue to myself.

2nd - according to documentation
https://docs.atlassian.com/software/jira/docs/api/REST/8.20.1/#user-findAssignableUsers
https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-user-search/#api-rest-api-2-user-assignable-search-get
parameter "query" is not always available . For example I'm using Jira 8.20.1 and "query"  doesn't work for me. 
If I use "username" I can find user which I'm looking for. It works  similar to "query" in newer version - can return multiple records.

3rd - Pagination doesn't work as expected for me. It looks like a some kind of protection against downloading the user list
/rest/api/2/user/assignable/search?project=PP&maxResults=1000&startAt=99
returns only 1 record even if there are more than 1000 users.

**Comment by jrich523** [comment]
@klaudiusz223 I tried to update the User struct in the way you have it but that didnt work for me. however Im not very good with Go so im not totally sure its correct.

But I see all the similar problems where it seems to only be able to query so many users and even the search feature doesnt help. I dont even make it out of A's. I wouldnt be surprised if we have over 10k users on our Jira platform.

On Prem v8.13.13

**Comment by jrich523** [comment]
@klaudiusz223 the curl worked and so does your branch!
Im mostly looking to use this for automation because we need a ticket for every PR,  so for those little one off tasks i wanted to do `jtb some description` which will create a ticket for me, and a git branch based on that ticket (jtb - jira ticket branch). the problem was assigning the ticket to me, and this fixes that, thanks!

**Comment by ankitpokhrel** [comment]
@klaudiusz223 @jrich523 Thank you for debugging the issue. Could you please try with the changes from #321 and see if it works for you? The PR also fixes the issue with assigning user with `jira issue edit`. You can either [compile the branch locally](https://github.com/ankitpokhrel/jira-cli#development) or install it with the following command. TIA

```
go install github.com/ankitpokhrel/jira-cli/cmd/jira@issue-198
```

**Comment by klaudiusz223** [comment]
@ankitpokhrel  Thanks for the patch
Unfortuantely it's not working for me. It's working partially in one aspect but i'm not using it in such way.

`jira --debug issue create --no-input -s"Summary"  -a example.username  ` -  doesn't work - task created but user not assigned correctly 

`jira issue assign ISSUE-1 example.username `  doesen't  work - user not assigned automaticaly but interactive mode for assigning user is started, unfortunately user list is narrowed to first 100 so I can not find proper user. If I choose myself from the list then I can succesfully assign task to myself. In previous version even this was not possible.

`jira issue assign $(jira me) `- interactive mode is started rather than immediately assign task to me

When I have more time I'll try to investigate it but I think I need to add a patch to include body in debug mode

**Comment by ankitpokhrel** [comment]
> jira issue assign ISSUE-1 example.username doesen't work - user not assigned automaticaly but interactive mode for assigning user is started

@klaudiusz223 interactive mode is started if there are multiple users with same username. The script cannot determine which user to choose if the jira API returns multiple results for the given username. 

Same is the case with `jira issue assign $(jira me)`, if searching your username returns multiple results, you will be presented with an interactive mode to select the user. Using `username` field to search for the user would solve the issue in this case but the field is deprecated and wont work everywhere (see error message below).

```sh
"errorMessages": [
    "The query parameter 'username' is not supported in GDPR strict mode."
  ]
```

Same is the case with this one `jira --debug issue create --no-input -s"Summary" -a example.username`, if you've email set in your profile try passing your email instead of `example.username` and it should work since email is unique (hopefully).

**Comment by ankitpokhrel** [comment]
The API seems to return best match at the top. We can introduce `--no-input` flag and that will take first result if multiple results are returned but that can still assign unintended user to the ticket.

**Comment by klaudiusz223** [comment]
If as a query parameter is used "query" it will not work for me no matter what i put in.  email address  or username. There is no such parameter in my version so it will not filter and jira will always retun first 99 users and me.
What You can check is to use "username" and "query" parameters in one query and see if You still has error regarding unsupported "username" parameter.

**Comment by ankitpokhrel** [comment]
@klaudiusz223 could you please mention Jira version you are using?

> What You can check is to use "username" and "query" parameters in one query and see if You still has error regarding unsupported "username" parameter.

The username field is deprecated and won't work in newer versions.

![Screen Shot 2022-03-08 at 9 04 15 PM](https://user-images.githubusercontent.com/2364546/157315980-28fe1498-f29d-4fb4-aed5-3ed3727b7a51.png)

**Comment by klaudiusz223** [comment]
As I wrote in one of my previous message it's 8.20.1
https://docs.atlassian.com/software/jira/docs/api/REST/8.20.1/#user-findAssignableUsers
I'm also aware that username field is deprecated as it is written in the 2nd link I provided in the same message.
I wonder what is the behaviour of jira if someone put in both of the parameters "query" and "username"
For example if I use "query" it is omitted w/o errors

**Comment by reidg44** [comment]
I'm dealing with the same issue as @klaudiusz223 on Jira Server v.8.20.x. Testing with the REST API show that swapping the "query" parameter to the "username" parameter fixes the issue for me.

Jira Server API ignores the "query" parameter altogether and just returns the first 100 users.

Just tested on jira-cli 1.0.0-beta but the issue persists.

**Comment by ankitpokhrel** [comment]
I ran into similar issue in one of the client's Jira installation. Will work on the fix. Its weird that this behaves differently in my local installation of v8.20.x 🤷‍♂️

**Comment by ankitpokhrel** [comment]
This should be fixed with #380

**Comment by h0adp0re** [comment]
Yes, can confirm. This works as expected now.

```sh
jira issue assign "TICKET-CODE" "$(jira me)"
```

Thanks a lot!
